dev/benchmarking/README.md

```diff
@@ -0,0 +1,305 @@
+The columns are listed in the same order as they are defined, to avoid having
+to explicitly name every colum in ingestion. The "id" column is left empty
+and will be automatically assigned on insert.
+
+To ingest the example CSV file from the command line,
+use the command below, substituting the 'PG_USER' username in the
+'.env' file for '&lt;username&gt;'.
+
+```shell
+CSV='examples/benchmark_result_example.csv' && \
+psql -U <username> -h localhost -p 5432 \
+ -c "\copy benchmark_result_view FROM '${CSV}' WITH (FORMAT csv, HEADER);"
+```
+
+#### Bulk ingestion using JSON
+
+To ingest the example JSON file using the psql client, use the command below.
+
+```
+\set content `cat examples/benchmark_example.json`
+SELECT ingest_benchmark_view(:'content'::jsonb);
+```
+
+### HTTP client
+
+The 'graphile/postgraphile' container provides an HTTP interface
+to the database via two url routes:
+
+- A GraphiQL page ([localhost:5000/graphiql][graphiql])
+  to aid visual exploration of the data model.
+  (The `--watch` flag on the command line. Not recommended for production.)
+- An endpoint that receives POST requests only (localhost:5000/graphql).
+
+#### Ingestion
+
+The script [graphql_submit.sh](./graphql_submit.sh) simplifies submission
+to the database via curl. Examples:
+
+```shell
+./graphql_submit.sh benchmarks examples/benchmark_example.json 
+./graphql_submit.sh results examples/benchmark_result_example.json
+```
+
+#### Querying
+
+The output of the query is a JSON object that is hard to read on the command line.
+Here is an example query in the shell:
+```shell
+curl -X POST \
+  -H "Content-Type: application/json"  \
+  --data '{"query": "{projectDetails{ projectName }}"}' \
+  localhost:5000/graphql
+```
+which yields
+```
+{"data":{"projectDetails":{"projectName":"Apache Arrow"}}}
+```
+
+Here is an example query using Python:
+```python
+>>> import json
+>>> import requests
+>>>
+>>> uri = "http://localhost:5000/graphql"
+>>> query = json.load(open("examples/graphql_query_environment_view.json"))
+>>> result = requests.post(uri, json=query)
+>>> message = "{benchmarkLanguage} {languageVersion} {dependencies}"
+>>>
+>>> for row in result.json()['data']['allEnvironmentViews']['edges']:
+...     print(message.format(**row['node']))
+...
+Python 2.7 {"six":"","numpy":"1.14","other_lib":"1.0"}
+Python 2.7 {"six":"","numpy":"1.15","other_lib":"1.0"}
+Python 3.6 {"boost":"1.42","numpy":"1.15"}
+```
+
+## Deployment
+
+(work in progress).
+
+
+## Quick reference
+
+- String variables `'have single quotes'`
+- Arrays `'{"have", "curly", "braces"}'::text[]` or `'{1, 2, 3}'::integer[]`
+- JSONb `'{"has":"this", "format":42}'::jsonb`
+- When comparing nullable values use `x IS DISTINCT FROM y` rather than `x = y`
+
+
+## Data model
+
+The primary view in the database is the
+[`benchmark_result_view`](#benchmark_result_view),
+which joins the actual benchmark result data with its context:
```

Comment:
it would be helpful here (perhaps as a follow-up task) to have a visual of the data-model (I don't know if our current doc-build supports it but one way would be to use [graphviz](https://www.sphinx-doc.org/en/master/usage/extensions/graphviz.html) to draw an [ER diagram](http://graphviz.996277.n3.nabble.com/Drawing-Database-ER-Diagrams-td1718.html) )

Reply:
Thank you for the suggestion—I am unfamiliar with graphviz but it looks do-able from the docs you linked to. Probably a follow-up task, but I will try to get something done quickly because seeing a diagram will be helpful for benchmark developers.

Reply:
There might be some autogenerate images from postgres as well, i'm just not familiar with them.

Reply:
Also the image you linked below is probably sufficient if there are easy instructions to regenerate it that would be great.

Reply:
Thank you! Sorry I didn't finish reading your other comments yet; i had to do some actual stuff for work but will get back to this Thursday night.

Reply:
Sadly the image was manually drawn. (using a free [ERDplus account](https://erdplus.com/#/))...will also explore how easy it is to generate text for this format though.